backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/backup/HybridBackupCommand.java

```diff
@@ -0,0 +1,304 @@
+        // which is not something we want to happen
+        setLock(null);
+
+        updateVmBackupPhase(VmBackupPhase.SUCCEEDED);
+        setSucceeded(true);
+    }
+
+    @Override
+    protected void endWithFailure() {
+        if (getParameters().getVmBackup().getPhase() == VmBackupPhase.REMOVING_SNAPSHOT) {
+            // TODO add audit log for possible leftover
+            endSuccessfully();
+        }
+
+        // Because the lock scope is Command endAction will reconstruct the lock and attempt to release it,
+        // which is not something we want to happen
+        setLock(null);
+
+        if (snapshotDao.exists(getVmId(), getParameters().getAutoGeneratedSnapshotId())) {
+            // Try to remove the snapshot
+            removeAutoGeneratedSnapshot(cloneContextAndDetachFromParent(), null, null);
+        }
+
+        updateVmBackupPhase(VmBackupPhase.FAILED);
+        setSucceeded(true);
+    }
+
+    @Override
+    public CommandCallback getCallback() {
+        return callbackProvider.get();
+    }
+
+    @Override
+    protected Map<String, Pair<String, String>> getExclusiveLocks() {
+        Map<String, Pair<String, String>> locks = new HashMap<>();
+        getDisks().forEach(disk -> locks.put(disk.getId().toString(),
+                LockMessagesMatchUtil.makeLockingPair(LockingGroup.DISK, EngineMessage.ACTION_TYPE_FAILED_DISK_IS_LOCKED)));
+        locks.put(getParameters().getVmBackup().getVmId().toString(),
+                LockMessagesMatchUtil.makeLockingPair(LockingGroup.VM, EngineMessage.ACTION_TYPE_FAILED_VM_IS_DURING_BACKUP));
+        return locks;
+
+    }
+
+    @Override
+    protected LockProperties applyLockProperties(LockProperties lockProperties) {
+        return lockProperties.withScope(LockProperties.Scope.Command);
+    }
+
+    @Override
+    public Map<String, String> getJobMessageProperties() {
+        if (jobProperties == null) {
+            jobProperties = super.getJobMessageProperties();
+            jobProperties.put("vm", getVmName());
+        }
+
+        return super.getJobMessageProperties();
+    }
+
+    private void auditLog(AuditLogType type) {
+        addCustomValue("VmName", getVm().getName());
+        addCustomValue("SnapshotName", StorageConstants.BACKUP_VM_AUTO_GENERATED_SNAPSHOT_DESCRIPTION);
+        auditLogDirector.log(this, type);
+    }
+
+    @Override
+    protected void restoreCommandState() {
+        Guid backupId = getParameters().getVmBackup().getId();
+        VmBackup vmBackup = vmBackupDao.get(backupId);
+        if (CollectionUtils.isEmpty(vmBackup.getDisks())) {
+            vmBackup.setDisks(getDisks());
+        }
+
+        getParameters().setVmBackup(vmBackup);
+    }
+
+    @Override
+    protected List<DiskImage> getDisks() {
+        return diskImageDao.getAllSnapshotsForVmSnapshot(getParameters().getAutoGeneratedSnapshotId());
+    }
+
+    @Override
+    public boolean reacquireLocks() {
```

Comment:
@ahadas haven't seen this done anywhere else, but seems to do the trick

Reply:
interesting, I didn't remember we have this one